api/src/main/java/bio/terra/tanagra/service/search/SqlVisitor.java

```diff
@@ -294,19 +296,89 @@ public String resolveTable(EntityVariable entityVariable) {
     private String resolveTable(Table table) {
       // `projectId.datasetId`.table
       return String.format(
           "`%s.%s`.%s", table.dataset().projectId(), table.dataset().datasetId(), table.name());
     }
 
+    /** Resolve a {@link Table} with a {@link TableFilter} into a SQL string. */
     private String resolveTable(Table table, @Nullable TableFilter tableFilter) {
       if (tableFilter == null) {
         return resolveTable(table);
+      } else if (tableFilter.arrayColumnFilter() != null) {
+        return resolveArrayColumnFilter(table, tableFilter.arrayColumnFilter(), false);
+      } else if (tableFilter.binaryColumnFilter() != null) {
+        return resolveBinaryColumnFilter(table, tableFilter.binaryColumnFilter(), false);
+      } else {
+        throw new IllegalArgumentException(
+            "Invalid table filter missing binary and array column filters.");
+      }
+    }
+
+    /**
+     * Resolve a {@link Table} with an {@link ArrayColumnFilter} into a SQL string.
+     *
+     * @param whereClauseOnly If true, only returns the part of the SQL string after the WHERE. This
+     *     is helpful when building an {@link ArrayColumnFilter} composed of multiple sub-filters.
+     */
+    private String resolveArrayColumnFilter(
+        Table table, ArrayColumnFilter arrayColumnFilter, boolean whereClauseOnly) {
+      String joinOperatorInWhereClause;
+      switch (arrayColumnFilter.operator()) {
+        case AND:
+          joinOperatorInWhereClause = "AND";
+          break;
+        case OR:
+          joinOperatorInWhereClause = "OR";
+          break;
+        default:
+          throw new IllegalArgumentException(
+              "Unknown array column filter operator type: " + arrayColumnFilter.operator());
+      }
+
+      // %s AND %s AND (%s OR %s) ...
+      StringBuilder joinedWhereClauses = new StringBuilder();
+      int subFiltersCtr = 0;
+      for (BinaryColumnFilter subFilter : arrayColumnFilter.binaryColumnFilters()) {
+        if (subFiltersCtr > 0) {
+          joinedWhereClauses.append(" " + joinOperatorInWhereClause + " ");
+        }
+        joinedWhereClauses.append(resolveBinaryColumnFilter(table, subFilter, true));
+        subFiltersCtr++;
+      }
+      for (ArrayColumnFilter subFilter : arrayColumnFilter.arrayColumnFilters()) {
+        if (subFiltersCtr > 0) {
+          joinedWhereClauses.append(" " + joinOperatorInWhereClause + " ");
+        }
+        joinedWhereClauses.append(resolveArrayColumnFilter(table, subFilter, true));
+        subFiltersCtr++;
+      }
+
+      if (whereClauseOnly) {
+        return "(" + joinedWhereClauses + ")";
+      } else {
+        // (SELECT * FROM `projectId.datasetId`.table WHERE joinedWhereClauses)
+        return String.format(
+            "(SELECT * FROM `%s.%s`.%s WHERE %s)",
+            table.dataset().projectId(),
+            table.dataset().datasetId(),
+            table.name(),
+            joinedWhereClauses);
```

Comment:
nit: maybe avoid the `whereClauseOnly` boolean, make this method only worry about the WHERE clause, and extract the full query logic out to the top level. This would also avoid this duplication across these two methods (and future filter methods)

Reply:
Good idea. Done.